src/resolver.rs

```diff
@@ -2381,7 +2381,23 @@ impl<'i> TypeAnnotator<'i> {
                         self.annotate(statement, StatementAnnotation::value(DATE_AND_TIME_TYPE));
                     }
                     AstLiteral::Real(value) => {
-                        self.annotate(statement, StatementAnnotation::value(get_real_type_name_for(value)));
+                        // If we have a float literal in an initializer (lhs) context, we need to see if the context expects a double or float type.
+                        // This is due to `get_real_type_name_for` always returning `REAL_TYPE` for values within f32 range which leads to incorrect typing
+                        // during initialization of double values.
+                        let Some(lhs_type) = ctx.lhs.as_ref().and_then(|lhs| {
+                            self.index
+                                .find_effective_type_by_name(lhs)
+                                .filter(|it| it.get_type_information().is_float())
+                        }) else {
+                            self.annotate(
+                                statement,
+                                StatementAnnotation::value(get_real_type_name_for(value)),
+                            );
+                            return;
+                        };
```

Comment:
Is `let type_name = ... else { ... };` followed by `self.annotate` an option here?